design-documents/database/declarative-schema-database-compatibility.md

```diff
@@ -0,0 +1,1093 @@
+     *
+     * Note that not all RDBMS support this, returning `null` indicates whether this is supported.
+     *
+     * @return string|null
+     */
+    public function getCharset(): ?string;
+}
+```
+Note: `AUTO_INCREMENT` is omitted intentionally. Auto increment must not be used for the application logic.
+
+#### Columns
+
+```
+class InformationSchema\Extra
+{
+    /**
+     * Returns the on-update clause of a column.
+     *
+     * @return string|null
+     */
+    public function getOnUpdate(): ?string {}
+}
+```
+
+```
+abstract class InformationSchema\Table\Column
+{
+    /**
+     * Defines the ANSI-SQL column type that is reflected by current class.
+     *
+     * Note that this does not neccessarily reflect the data type name of all RDBMS (e.g. PostgreSQL uses `text` instead
+     * of `CLOB`).
+     * Thus the value stored herein should be used to map ANSI-SQL data types to RDBMS specific datatypes. 
+     */
+    public const TYPE = '';
+
+    /**
+     * Retuns the name of the table to which colum belongs.
+     *
+     * @return string
+     */
+    public function getTableName(): string {}
+
+    /**
+     * Returns the name of the column.
+     *
+     * @return string
+     */
+    public function getName(): string {}
+
+    /**
+     * Returns whether column is nullable.
+     *
+     * @return bool
+     */
+    public function getIsNullable(): bool {}
+
+    /**
+     * Returns the data object for the extra information of a column.
+     *
+     * Note that not all RDBMS support this, returning `null` indicates whether this is supported.
+     *
+     * @return ColumnExtraInterface|null
+     */
+    public function getExtra(): ?ColumnExtraInterface {}
+
+    /**
+     * Returns the comment included in the column definition.
+     *
+     * Note that not all RDBMS support this, returning `null` indicates whether this is supported.
+     *
+     * @return string|null
+     */
+    public function getComment(): ?string {}
+
+    /**
+     * Returns the default value of the column.
+     *
+     * @return mixed
+     */
+    public function getDefaultValue() {}
+}
+```
+
+##### Boolean types
+
+```
+class InformationSchema\Table\BoolColumn extends Column
+{
+    public const TYPE = 'BOOLEAN';
+
+    /**
+     * Returns the default value of the column.
+     *
+     * @return int|null
+     */
+    public function getDefaultValue(): ?int {}
+}
+```
+Note that MySQL / MariaDB maps this type to `tinyint(1)` thus resulting in an integer type.
+
+##### Numeric types
+
+```
+class InformationSchema\Table\BitColumn extends Column
+{
+    public const TYPE = 'BIT';
+
+    /**
+     * Returns the amount of bits the column can store.
+     *
+     * @return int
+     */
+    public function getBitLength(): int {}
+
+    /**
+     * Returns the default value of the column.
+     *
+     * @return int|null
+     */
+    public function getDefaultValue(): ?int {}
+}
+```
+
+```
+class InformationSchema\Table\VarBitColumn extends Column
```

Comment:
Do we need to introduce types that are not used by application currently? We will have to make sure that these types support across different databases engines.

Reply:
@melnikovi No they are not necessarily needed, I just added them for the sake of being complete. Technically both `bit` and `bit varying` have been dropped as of SQL-2003, so we can stimply ommit these.

Reply:
Yes, let's omit them.

Reply:
Ok, I removed the classes for `bit` and `bit varying` from the proposal.